Support Cassandra CDC logical types (new option decodeCDCDataTypes)

aymkhalil opened the pull request:
This patch will handle avro logical types coming from an upstream CSC. The following C* types will be handled:
* BigDecimal  #29 
* BigInteger
* CqlDuration
* UUID (nothing special here, just a string that already worked).

The implementation relies on the ConvertingCodecs from C* (StringTo[BigDecimal|BigInteger|CqlDuration] ConvertingCodecs.

I will follow up with a patch with a flag to disable this conversion from non-CSC users.

eolivelli commented on pulsar-impl/src/main/java/com/datastax/oss/sink/pulsar/AvroTypeUtil.java:
```diff
@@ -15,24 +15,74 @@
   }
+
+  private static boolean isVarint(Object varint) {
+    return varint instanceof ByteBuffer;
```
I am not sure that this assumption always holds.

eolivelli commented on pulsar-impl/src/main/java/com/datastax/oss/sink/pulsar/AvroTypeUtil.java:
```diff
@@ -15,24 +15,74 @@
+    return false;
+  }
+
+  public static Object handleLogicalType(Object value) {
```
I would call this "handleCassandraCDCLogicalType", this way it is a hard reference to CDC
adding a link to the docs of CDC would help as well probably

eolivelli commented on pulsar-impl/src/main/java/com/datastax/oss/sink/pulsar/AvroTypeUtil.java:
```diff
@@ -15,24 +15,74 @@
     return record != null && record.getSchemaType() == SchemaType.AVRO && isMapOrList(fieldValue);
   }
 
+  public static boolean shouldHandleLogicalType(GenericRecord record, Object fieldValue) {
```
what about "isCassandraCDCLogicalType" ?

eolivelli commented on pulsar-impl/src/main/java/com/datastax/oss/sink/pulsar/AvroTypeUtil.java:
```diff
@@ -15,24 +15,74 @@
+    if (record == null && record.getSchemaType() != SchemaType.AVRO) {
+      return false;
+    }
+    if (isVarint(fieldValue)) {
```
we should have access to the schema somehow here
we should use the Schema to safely decode these objects

eolivelli reviewed (changes requested):
We are on our way.

I believe that we should use the Schema (we should have it) to detect varints.
Otherwise even if you receive a "bytes" datatype (anything that AVRO decodes as a ByteBuffer) you would interpret it as "varint"

aymkhalil commented:
In the latest path:
* I changed the logical type check to respect the schema and avoid reflection
* Update the schemas in the test to use optional logical type schema (i.e. Union with null) to match the upstream CDC
* Added a `decodeCDCDataTypes` flag so the feature is turned of altogether for existing users to reduce regression chances

eolivelli commented on pulsar-impl/src/main/java/com/datastax/oss/sink/pulsar/AvroTypeUtil.java:
```diff
@@ -15,24 +15,131 @@
 
-  private static final Logger log = LoggerFactory.getLogger(AvroTypeUtil.class);
+  private static Map<String, Conversion<?>> logicalTypeConverters = new HashMap<>();
+  private static boolean decodeCDCDataTypes;
```
we need something to reset this value.

In production there is no big problem, because every Sink runs into its own Classloader, so multiple Sinks inside the same JVM are isolated.

but probably in unit tests this is a problem

aymkhalil replied:
good point. will do. the only caveat is I cannot deregister the Avro types on org.apache.avro.LogicalTypes because there is no "deregister api" - but I still have control over the flag on our side...

eolivelli reviewed (approved):
Great work !

LGTM

I have left one comment about tests
